Adopt more spans in WebAudio code

cdumez opened the pull request:
#### 4692bb90e0930842c44c92be11a311d69ae09add
<pre>
Adopt more spans in WebAudio code
<a href="https://bugs.webkit.org/show_bug.cgi?id=284704">https://bugs.webkit.org/show_bug.cgi?id=284704</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::singleElementSpan):
* Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
(WebCore::AudioBuffer::rawChannelData):
(WebCore::AudioBuffer::copyTo const):
* Source/WebCore/Modules/webaudio/AudioBuffer.h:
* Source/WebCore/Modules/webaudio/AudioListener.cpp:
(WebCore::AudioListener::updateValuesIfNeeded):
* Source/WebCore/Modules/webaudio/AudioParam.cpp:
(WebCore::replaceNaNValues):
(WebCore::AudioParam::finalValue):
(WebCore::AudioParam::calculateSampleAccurateValues):
(WebCore::AudioParam::calculateFinalValues):
(WebCore::AudioParam::calculateTimelineValues):
* Source/WebCore/Modules/webaudio/AudioParam.h:
* Source/WebCore/Modules/webaudio/AudioParamTimeline.cpp:
(WebCore::AudioParamTimeline::valueForContextTime):
(WebCore::AudioParamTimeline::valuesForFrameRange):
(WebCore::AudioParamTimeline::valuesForFrameRangeImpl):
* Source/WebCore/Modules/webaudio/AudioParamTimeline.h:
* Source/WebCore/Modules/webaudio/AudioWorkletNode.cpp:
(WebCore::AudioWorkletNode::process):
* Source/WebCore/Modules/webaudio/BiquadDSPKernel.cpp:
(WebCore::hasConstantValue):
(WebCore::BiquadDSPKernel::updateCoefficientsIfNecessary):
(WebCore::BiquadDSPKernel::updateCoefficients):
* Source/WebCore/Modules/webaudio/BiquadDSPKernel.h:
* Source/WebCore/Modules/webaudio/BiquadProcessor.cpp:
(WebCore::BiquadProcessor::processOnlyAudioParams):
(WebCore::BiquadProcessor::getFrequencyResponse):
* Source/WebCore/Modules/webaudio/ConstantSourceNode.cpp:
(WebCore::ConstantSourceNode::process):
* Source/WebCore/Modules/webaudio/ConvolverNode.cpp:
(WebCore::ConvolverNode::setBufferForBindings):
* Source/WebCore/Modules/webaudio/DelayDSPKernel.cpp:
(WebCore::DelayDSPKernel::processARate):
(WebCore::DelayDSPKernel::processOnlyAudioParams):
* Source/WebCore/Modules/webaudio/DynamicsCompressorNode.cpp:
(WebCore::DynamicsCompressorNode::processOnlyAudioParams):
* Source/WebCore/Modules/webaudio/GainNode.cpp:
(WebCore::GainNode::process):
(WebCore::GainNode::processOnlyAudioParams):
* Source/WebCore/Modules/webaudio/OscillatorNode.cpp:
(WebCore::clampFrequency):
(WebCore::OscillatorNode::calculateSampleAccuratePhaseIncrements):
(WebCore::OscillatorNode::processKRate):
(WebCore::OscillatorNode::process):
* Source/WebCore/Modules/webaudio/PannerNode.cpp:
(WebCore::PannerNode::processOnlyAudioParams):
(WebCore::PannerNode::processSampleAccurateValues):
* Source/WebCore/Modules/webaudio/ScriptProcessorNode.cpp:
(WebCore::ScriptProcessorNode::process):
* Source/WebCore/Modules/webaudio/StereoPannerNode.cpp:
(WebCore::StereoPannerNode::process):
(WebCore::StereoPannerNode::processOnlyAudioParams):
* Source/WebCore/platform/audio/AudioBus.cpp:
(WebCore::AudioBus::AudioBus):
(WebCore::AudioBus::setChannelMemory):
(WebCore::AudioBus::copyWithSampleAccurateGainValuesFrom):
* Source/WebCore/platform/audio/AudioBus.h:
* Source/WebCore/platform/audio/AudioChannel.h:
* Source/WebCore/platform/audio/AudioResampler.cpp:
(WebCore::AudioResampler::process):
* Source/WebCore/platform/audio/AudioResamplerKernel.cpp:
(WebCore::AudioResamplerKernel::getSourceSpan):
(WebCore::AudioResamplerKernel::getSourcePointer): Deleted.
* Source/WebCore/platform/audio/AudioResamplerKernel.h:
* Source/WebCore/platform/audio/StereoPanner.cpp:
(WebCore::StereoPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/StereoPanner.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::render):

Canonical link: <a href="https://commits.webkit.org/287880@main">https://commits.webkit.org/287880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4752eddf1ec3e6317f83352633653a4f9765e263

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81181 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35124 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85710 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32167 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83291 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/724 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8516 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63381 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21152 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84250 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/472 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73886 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43679 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/369 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28048 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30625 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71889 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28615 "Found 1 new test failure: fast/attachment/cocoa/wide-attachment-rendering.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87145 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8411 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5971 "Found 1 new test failure: fast/attachment/cocoa/wide-attachment-rendering.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71688 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8590 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69721 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70923 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17648 "Found 1 new test failure: fast/attachment/cocoa/wide-attachment-rendering.html (failure)") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14970 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13889 "Found 1 new test failure: fast/attachment/cocoa/wide-attachment-rendering.html (failure)") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12584 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8372 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13896 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8209 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11729 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10017 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/Modules/webaudio/AudioParam.cpp:
```diff
@@ -270,23 +272,23 @@ bool AudioParam::hasSampleAccurateValues() const
+void AudioParam::calculateSampleAccurateValues(std::span<float> values)
 {
-    bool isSafe = context() && context()->isAudioThread() && values && numberOfValues;
+    bool isSafe = context() && context()->isAudioThread() && values.data() && !values.empty();
```
Do we need the `values.data()` check?

darinadler commented on Source/WebCore/Modules/webaudio/AudioParam.cpp:
```diff
@@ -270,23 +272,23 @@ bool AudioParam::hasSampleAccurateValues() const
+void AudioParam::calculateFinalValues(std::span<float> values, bool sampleAccurate)
 {
-    bool isGood = context() && context()->isAudioThread() && values && numberOfValues;
+    bool isGood = context() && context()->isAudioThread() && values.data() && !values.empty();
```
Do we need the `values.data()` check?

darinadler commented on Source/WebCore/Modules/webaudio/AudioParam.cpp:
```diff
@@ -328,17 +330,17 @@ void AudioParam::calculateFinalValues(float* values, unsigned numberOfValues, bo
 
     // Clamp values based on range allowed by AudioParam's min and max values.
-    VectorMath::clamp(values, minValue(), maxValue(), values, numberOfValues);
+    VectorMath::clamp(values.data(), minValue(), maxValue(), values.data(), values.size());
```
Seems like this needs to change to take a span. I'm sure you noticed.

cdumez replied:
Yes, I've decided to keep VectorMath out of this PR to keep the patch size reasonable.

darinadler commented on Source/WebCore/Modules/webaudio/AudioParamTimeline.cpp:
```diff
@@ -448,17 +448,17 @@ float AudioParamTimeline::valuesForFrameRangeImpl(size_t startFrame, size_t endF
         ASSERT(fillToEndFrame >= startFrame);
-        unsigned fillToFrame = static_cast<unsigned>(fillToEndFrame - startFrame);
-        fillToFrame = std::min(fillToFrame, numberOfValues);
+        size_t fillToFrame = static_cast<size_t>(fillToEndFrame - startFrame);
```
Do we need the `static_cast`? I think it has no effect.

darinadler commented on Source/WebCore/Modules/webaudio/AudioParamTimeline.cpp:
```diff
@@ -448,17 +448,17 @@ float AudioParamTimeline::valuesForFrameRangeImpl(size_t startFrame, size_t endF
 
         const AutomationState currentState = {
-            numberOfValues,
+            static_cast<unsigned>(values.size()),
```
What guarantees this doesn't truncate? Can we use something that checks?

darinadler commented on Source/WebCore/Modules/webaudio/AudioParamTimeline.cpp:
```diff
@@ -448,17 +448,17 @@ float AudioParamTimeline::valuesForFrameRangeImpl(size_t startFrame, size_t endF
             controlRate,
             samplingPeriod,
-            fillToFrame,
+            static_cast<unsigned>(fillToFrame),
```
What guarantees this doesn't truncate? Can we use something that checks?

darinadler commented on Source/WebCore/platform/audio/AudioResampler.cpp:
```diff
@@ -87,12 +87,12 @@ void AudioResampler::process(AudioSourceProvider* provider, AudioBus* destinatio
-        ASSERT(fillPointer);
-        if (!fillPointer)
+        std::span<float> fillSpan = m_kernels[i]->getSourceSpan(framesToProcess, &framesNeeded);
+        ASSERT(fillSpan.data());
```
Should this assertion come after the `empty()` check?

cdumez committed:
Adopt more spans in WebAudio code
https://bugs.webkit.org/show_bug.cgi?id=284704

Reviewed by Darin Adler.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::singleElementSpan):
* Source/WebCore/Modules/webaudio/AudioBuffer.cpp:
(WebCore::AudioBuffer::rawChannelData):
(WebCore::AudioBuffer::copyTo const):
* Source/WebCore/Modules/webaudio/AudioBuffer.h:
* Source/WebCore/Modules/webaudio/AudioListener.cpp:
(WebCore::AudioListener::updateValuesIfNeeded):
* Source/WebCore/Modules/webaudio/AudioParam.cpp:
(WebCore::replaceNaNValues):
(WebCore::AudioParam::finalValue):
(WebCore::AudioParam::calculateSampleAccurateValues):
(WebCore::AudioParam::calculateFinalValues):
(WebCore::AudioParam::calculateTimelineValues):
* Source/WebCore/Modules/webaudio/AudioParam.h:
* Source/WebCore/Modules/webaudio/AudioParamTimeline.cpp:
(WebCore::AudioParamTimeline::valueForContextTime):
(WebCore::AudioParamTimeline::valuesForFrameRange):
(WebCore::AudioParamTimeline::valuesForFrameRangeImpl):
* Source/WebCore/Modules/webaudio/AudioParamTimeline.h:
* Source/WebCore/Modules/webaudio/AudioWorkletNode.cpp:
(WebCore::AudioWorkletNode::process):
* Source/WebCore/Modules/webaudio/BiquadDSPKernel.cpp:
(WebCore::hasConstantValue):
(WebCore::BiquadDSPKernel::updateCoefficientsIfNecessary):
(WebCore::BiquadDSPKernel::updateCoefficients):
* Source/WebCore/Modules/webaudio/BiquadDSPKernel.h:
* Source/WebCore/Modules/webaudio/BiquadProcessor.cpp:
(WebCore::BiquadProcessor::processOnlyAudioParams):
(WebCore::BiquadProcessor::getFrequencyResponse):
* Source/WebCore/Modules/webaudio/ConstantSourceNode.cpp:
(WebCore::ConstantSourceNode::process):
* Source/WebCore/Modules/webaudio/ConvolverNode.cpp:
(WebCore::ConvolverNode::setBufferForBindings):
* Source/WebCore/Modules/webaudio/DelayDSPKernel.cpp:
(WebCore::DelayDSPKernel::processARate):
(WebCore::DelayDSPKernel::processOnlyAudioParams):
* Source/WebCore/Modules/webaudio/DynamicsCompressorNode.cpp:
(WebCore::DynamicsCompressorNode::processOnlyAudioParams):
* Source/WebCore/Modules/webaudio/GainNode.cpp:
(WebCore::GainNode::process):
(WebCore::GainNode::processOnlyAudioParams):
* Source/WebCore/Modules/webaudio/OscillatorNode.cpp:
(WebCore::clampFrequency):
(WebCore::OscillatorNode::calculateSampleAccuratePhaseIncrements):
(WebCore::OscillatorNode::processKRate):
(WebCore::OscillatorNode::process):
* Source/WebCore/Modules/webaudio/PannerNode.cpp:
(WebCore::PannerNode::processOnlyAudioParams):
(WebCore::PannerNode::processSampleAccurateValues):
* Source/WebCore/Modules/webaudio/ScriptProcessorNode.cpp:
(WebCore::ScriptProcessorNode::process):
* Source/WebCore/Modules/webaudio/StereoPannerNode.cpp:
(WebCore::StereoPannerNode::process):
(WebCore::StereoPannerNode::processOnlyAudioParams):
* Source/WebCore/platform/audio/AudioBus.cpp:
(WebCore::AudioBus::AudioBus):
(WebCore::AudioBus::setChannelMemory):
(WebCore::AudioBus::copyWithSampleAccurateGainValuesFrom):
* Source/WebCore/platform/audio/AudioBus.h:
* Source/WebCore/platform/audio/AudioChannel.h:
* Source/WebCore/platform/audio/AudioResampler.cpp:
(WebCore::AudioResampler::process):
* Source/WebCore/platform/audio/AudioResamplerKernel.cpp:
(WebCore::AudioResamplerKernel::getSourceSpan):
(WebCore::AudioResamplerKernel::getSourcePointer): Deleted.
* Source/WebCore/platform/audio/AudioResamplerKernel.h:
* Source/WebCore/platform/audio/StereoPanner.cpp:
(WebCore::StereoPanner::panWithSampleAccurateValues):
* Source/WebCore/platform/audio/StereoPanner.h:
* Source/WebCore/platform/audio/cocoa/AudioDestinationCocoa.cpp:
(WebCore::AudioDestinationCocoa::render):

Canonical link: https://commits.webkit.org/287880@main
